docs/proposals/rebalancing-clusters-across-shards-dynamically.md

```diff
@@ -0,0 +1,144 @@
+
+The service account used by the application controller has read access to all the resources in the cluster. Thus, instead of setting the environment variable ARGOCD_CONTROLLER_REPLICAS representing the number of replicas, the number of replicas can be read directly from the number of healthy replicas of the application controller deployment.
+
+We would assign one application controller to manage one of the shards and we will store this assignment list in Redis. The mapping of Application Controller to Shard will store the below information:
+
+* Name/Id of the shard
+* Name of the Application Controller currently managing the shard
+* Last time of successful update to Redis (Heartbeat)
+
+The mapping will be updated in Redis every X (heartbeat interval) seconds with the help of heartbeat process performed by every application controller. If the heartbeat was not performed by the application controller for a certain time, the application controller is assumed to be unhealthy and the number of healthy/managed shards would be reduced, that is, the number of healthy replicas of the application controller deployment changes.
+
+The heartbeat interval will be a configurable parameter initialized while setting up the application controller. This way, users will be able to control the frequency at which they want the heartbeat process to take place.
+
+Anytime the number of healthy replicas of application controllers is different from the number of application controllers to shard mappings, we would re-distribute the clusters among the healthy replicas again. We can summarize the above statement using the below formula:
+
+```
+Number of Replicas ≠ Count of {Application Controller, Shard} mapping
+```
+
+The below logic can be used to perform application controller to shard assignment:
+
+1) If a new application controller is added, that is, a new shard is added, we would perform the re-distribution of clusters among the shards with the existing sharding algorithm being used.
+
+2) In scenarios when one of the application controllers is identified to be unhealthy, the rest of the application controllers will re-distribute the load(clusters) managed by the application controller and start managing the orphaned applications. 
+
+How will this work? 
+* The application controller will query Redis for the status of all the application controllers and last updated heartbeat timestamp.
+* It will check if any application controller is flagged as Unhealthy or has not updated its status in Redis during the heartbeat process for a certain period of time.
+* If the status for an application controller was already flagged as Unhealthy, it will pick one of the clusters from its list and assign the cluster to itself.
```

Comment:
I am afraid that this approach will lead to race conditions issues. Maybe this should be done by all applications controllers but having just one dealing with analyzing heartbeat and redistributing clusters.

Reply:
Wouldn't you then have the issue that if that controller goes down you no longer have a mechanism to perform the healthcheck, so the health check controller never gets marked as unhealthy?

Reply:
I agree with @rouke-broersma here.. Relying on a single controller has a bunch of issues as well, for example, if the controller performing the redistribution goes down or starts to distribute in a faulty manner.

Reply:
Here I agree with @leoluz about race conditions. Controller health might be flapping and it is hard to predict what happens sooner: controller instance recovers or the shards are redistributed among healthy controllers.

I'm proposing to just don't detect unhealthy clusters and let kubenetes/operators deal with it:

* healthy controllers just "claim" shards and update ConfigMap
* each controller should have a readiness probe that returns healthy only if the controller successfully claimed shard and updated heartbeat within the allowed timeframe.

So if one controller instance is unhealthy one shard won't be processed but operator will know about it because readiness probe would be failing.

Reply:
Thanks @alexmt, the suggestion of not detecting unhealthy shards does make sense.. In this case, we would need to re-introduce the liveness probe or update the readiness probe with the condition where we check whether the controller was able to claim a shard or not.